### PR TITLE
feat: new component Sheet & PromoSheet

### PR DESCRIPTION
**Pull request by Avol-V**
Add components `Sheet` (`MobileModal`) and `PromoSheet`.

### CODE REVIEW

File: src/components/PromoSheet/PromoSheet.tsx [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        };
+
+        image.src = imageSrc;
+    }, [imageSrc]);
+
+    return (
+        <Sheet
+            className={cn(null, className)}
+            contentClassName={cn('content', contentClassName)}
+            visible={visible && loaded}
```
Comment (korvin89) [inline src/components/PromoSheet/PromoSheet.tsx:101]: Can we have a lag here in case of slow connection?

File: src/components/Sheet/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+## MobileModal
+
+Sheet component for mobile devices
+
+### PropTypes
+
+| Property                 | Type       | Required | Default     | Description                                                                                                                                                                 |
+| :----------------------- | :--------- | :------: | :---------- | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| visible                  | `boolean`  |    ✓     |             | Show/hide sheet                                                                                                                                                             |
```
Comment (korvin89) [inline src/components/Sheet/README.md:9]: Why it should be a required property?

**Review by korvin89 — COMMENTED** [review]
You could add [codeowners](https://github.com/gravity-ui/uikit/blob/main/CODEOWNERS) right here. Sheet - @korvin89 PromoSheet - @Avol-V

File: src/components/PromoSheet/PromoSheet.tsx (hunk 1 above)
Comment (Avol-V) [reply src/components/PromoSheet/PromoSheet.tsx:101]: It's a promo-informer and should be visible only when it's ready to show. So it's just hidden while image loading and we don't need any visible loader.

File: src/components/Sheet/README.md (hunk 2 above)
Comment (Avol-V) [reply src/components/Sheet/README.md:9]: It was required in the original component. Maybe to highlight that you should to specify `visible` property to control visibility of the component, it's useless without `visible` property.

**Review by korvin89 — APPROVED** [review]

File: src/components/Sheet/Sheet.scss [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        line-height: 28px;
+        text-align: center;
+        overflow: hidden;
+        white-space: nowrap;
+        text-overflow: ellipsis;
+    }
+}
+
+.yc-root {
+    --yc-sheet-content-paddings: 0 10px;
```
Comment (amje) [inline src/components/Sheet/Sheet.scss:100]: `padding`, in singular

File: src/components/Sheet/Sheet.tsx [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+import ReactDOM from 'react-dom';
+import {SheetContentContainer} from './SheetContent';
+import {sheetBlock} from './constants';
+
+import './Sheet.scss';
+
+export interface SheetProps {
+    onClose?: () => void;
+    /** Show/hide sheet */
+    visible: boolean;
```
Comment (amje) [inline src/components/Sheet/Sheet.tsx:11]: We use `open` for this

File: src/components/Sheet/Sheet.tsx [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    /** Class name for the sheet window */
+    className?: string;
+    /** Class name for the sheet content */
+    contentClassName?: string;
+    /** Class name for the swipe area */
+    swipeAreaClassName?: string;
+    /** Enable the behavior in which you can close the sheet window with a swipe down if the content is scrolled to its top (`contentNode.scrollTop === 0`) or has no scroll at all */
+    allowHideOnContentScroll?: boolean;
+    /** Hide top bar with resize handle */
+    noTopBar?: boolean;
```
Comment (amje) [inline src/components/Sheet/Sheet.tsx:25]: `hideTopBar`

File: src/components/Sheet/SheetContent.tsx [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+const ACCELERATION_Y_MIN = -0.02;
+
+let hashHistory: string[] = [];
+
+type Status = 'showing' | 'hiding';
+
+interface SheetContentBaseProps {
+    hideSheet: () => void;
+    content: React.ReactNode;
+    visible: boolean;
```
Comment (amje) [inline src/components/Sheet/SheetContent.tsx:20]: `open`

File: src/components/Sheet/SheetContent.tsx [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+const ACCELERATION_Y_MAX = 0.08;
+const ACCELERATION_Y_MIN = -0.02;
+
+let hashHistory: string[] = [];
+
+type Status = 'showing' | 'hiding';
+
+interface SheetContentBaseProps {
+    hideSheet: () => void;
+    content: React.ReactNode;
```
Comment (amje) [inline src/components/Sheet/SheetContent.tsx:19]: Why don't use children for it?

File: src/components/Sheet/SheetContent.tsx [hunk 8]
```diff
@@ -0,0 +1,482 @@
...
+        return (
+            platform !== Platform.BROWSER &&
+            history.action === 'POP' &&
+            prevProps.location.hash !== location.hash &&
+            location.hash !== `#${id}`
+        );
+    }
+}
+
+function withRouterWrapper(Component: React.ComponentType<SheetContentInnerProps>) {
```
Comment (amje) [inline src/components/Sheet/SheetContent.tsx:472]: Work with router/urls should be outside of the library. This component should be "dumb"

File: src/components/Sheet/Sheet.scss [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    --yc-sheet-transition-duration: 0.3s;
```
Comment (amje) [inline src/components/Sheet/Sheet.scss:101]: Will the component allow style overwrites? Why we define these?

File: src/components/Sheet/Sheet.tsx [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+interface SheetState {
+    visible: boolean;
+}
+
+export class Sheet extends React.Component<SheetProps, SheetState> {
```
Comment (amje) [inline src/components/Sheet/Sheet.tsx:32]: Why class component, not function?

File: src/components/Sheet/Sheet.tsx [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    private static bodyScrollLocksCount = 0;
+    private static bodyInitialOverflow: string | undefined = undefined;
+
+    static lockBodyScroll() {
```
Comment (amje) [inline src/components/Sheet/Sheet.tsx:36]: Its already implemented in `useBodyScrollLock`

File: src/components/Sheet/Sheet.tsx [hunk 12]
```diff
@@ -0,0 +1,142 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import {SheetContentContainer} from './SheetContent';
```
Comment (amje) [inline src/components/Sheet/Sheet.tsx:3]: Let's name member as the filename

File: src/components/Sheet/SheetContent.tsx [hunk 13]
```diff
@@ -0,0 +1,482 @@
...
+    deltaY: number;
+    prevInnerContentHeight: number;
+    swipeAreaTouched: boolean;
+    contentTouched: boolean;
+    veilTouched: boolean;
+    isAnimating: boolean;
+    inWindowResizeScope: boolean;
+}
+
+class SheetContent extends React.Component<SheetContentInnerProps, SheetContentState> {
```
Comment (amje) [inline src/components/Sheet/SheetContent.tsx:56]: Can be also implemented as a function

File: src/components/PromoSheet/PromoSheet.scss [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+.yc-root {
+    --yc-promo-sheet-margin: 8px;
+    --yc-promo-sheet-padding: 20px;
+    --yc-promo-sheet-border-radius: 12px;
+    --yc-promo-sheet-header-margin: 12px;
+    --yc-promo-sheet-message-margin: 16px;
+    --yc-promo-sheet-image-margin: 12px;
+    --yc-promo-sheet-foreground: var(--yc-color-text-light-primary);
+    --yc-promo-sheet-background: var(--yc-my-color-brand-normal);
```
Comment (amje) [inline src/components/PromoSheet/PromoSheet.scss:67]: Same question here, we would allow to style this component?

**Review by amje — CHANGES_REQUESTED** [review]

File: src/components/PromoSheet/PromoSheet.scss (hunk 14 above)
Comment (Avol-V) [reply src/components/PromoSheet/PromoSheet.scss:67]: Yes

**Review by amje — APPROVED** [review]
Components will be refactored in the future #435

**Pull request merged**